blockchain/chain/src/store/tipset_tracker.rs

```diff
@@ -0,0 +1,83 @@
+use std::{collections::HashMap, sync::Arc};
+
+use blocks::{BlockHeader, Tipset};
+use cid::Cid;
+use clock::ChainEpoch;
+use ipld_blockstore::BlockStore;
+
+use super::Error;
+
+/// Tracks blocks by their height for the purpose of forming tipsets.
+#[derive(Default)]
+pub struct TipsetTracker<DB> {
+    entries: HashMap<ChainEpoch, Vec<Cid>>,
+    db: Arc<DB>,
+}
+
+impl<DB: BlockStore> TipsetTracker<DB> {
+    pub fn new(db: Arc<DB>) -> Self {
+        Self {
+            entries: HashMap::new(),
+            db,
+        }
+    }
+
+    /// Adds a block header to the tracker.
+    pub fn add(&mut self, header: &BlockHeader) {
+        let entries = self.entries.entry(header.epoch()).or_default();
+
+        for cid in entries.iter() {
+            if cid == header.cid() {
+                log::debug!("tried to add block to tipset tracker that was already there");
+                return;
+            }
+
+            if let Ok(Some(block)) = self.db.get::<BlockHeader>(cid) {
+                if header.miner_address() == block.miner_address() {
+                    log::warn!(
+                        "Have multiple blocks from miner {} at height {} in our tipset cache {}-{}",
+                        header.miner_address(),
+                        header.epoch(),
+                        header.cid(),
+                        cid
+                    );
+                }
+            }
+        }
```

Comment:
This seems like it might be worth to split up these checks into two loops, I would think it might be nice to short circuit the cids checks, so if any are the same, you don't have to load and deserialize a whole block header from storage. 

The second check also seems like a good candidate for a cache in the future (no need to do now) but may be worth it to cache miner addresses -> block cids, since entries are never removed? What are your thoughts?

Reply:
Heh, I started out caching miner addresses to avoid having to fetch the entire block header from the blockstore. In the end I reverted to this because I don't know how crucial it is for this tracker to minimize its memory footprint.

The `expand` method could potentially benefit from caching the parent tipset keys as well in order to avoid fetching block headers with different parents, but again, I'm unsure about the memory/speedup trade-offs.

Reply:
Yeah, let's leave it as is for now, but something for us to keep in mind